src/onINP.ts

```diff
@@ -57,20 +63,88 @@ const getInteractionCountForNavigation = () => {
 let longestInteractionList: Interaction[] = [];
 
 // A mapping of longest interactions by their interaction ID.
 // This is used for faster lookup.
-const longestInteractionMap: {[interactionId: string]: Interaction} = {};
+const longestInteractionMap: {[interactionId: string]: Interaction} =
+  Object.create(null);
+
+// A mapping between a particular frame's render time and all of the
+// event timing entries that occurred within that frame. Since event timing
+// entries round their duration to the nearest 8sm, this ends up being
+// a best-effort guess, but but debugging and attribution practices it's
+// useful to see all events the occurred within the same frame as a particular
+// interaction, so it's worth doing the best-effort guess.
+// const pendingEntriesGroupMap: Map<number, PendingEntriesGroup> = new Map();
+const pendingEntriesGroupMap: {[renderTime: string]: PendingEntriesGroup} =
+  Object.create(null);
+
+/**
+ * Gets the "frame-normalized" render time for the given entry.
+ * In other words, all entries passed to this function that were processed
+ * within the same frame should return the same render time. This allows you
+ * to group by this render time value.
+ * This function works by referencing `pendingEntriesGroupMap` and using
+ * an existing render time if one is found (otherwise creating a new one).
+ */
+const groupEntryByRenderTime = (entry: PerformanceEventTiming) => {
+  // Since event timing entry `duration` values are rounded to the nearest
+  // 8ms, it's possible that `startTime + duration` could be less than the
+  // `processingEnd` time, so we cap the render time there since render times
+  // must be after processing.
+  const renderTime = Math.max(
+    entry.startTime + entry.duration,
+    entry.processingEnd,
+  );
+
+  let matchingRenderTime;
+
+  for (const key in pendingEntriesGroupMap) {
+    const prevRenderTime = Number(key);
+    const entriesGroup = pendingEntriesGroupMap[prevRenderTime];
+
+    // If a previous render time is within 8ms of the current render time,
+    // assume they were part of the same frame and re-use the previous time.
+    // Also break out of the loop because all subsequent times will be newer.
+    if (Math.abs(renderTime - prevRenderTime) <= 8) {
+      matchingRenderTime = prevRenderTime;
+      entriesGroup.entries.push(entry);
+      break;
+    }
+
+    // If a previous render time is more than 2 seconds before the current
+    // render time, assume all entries within that frame have been fully
+    // processed and remove it from the pending frames map (to free up space).
+    // NOTE: this is not a perfect heuristic, but if it's wrong, the worst
+    // thing that will happen is some entries may not get reported along
+    // with INP (but that is extremely unlikely). It will not impact
+    // the accuracy of the INP value that is reported.
+    if (prevRenderTime + 2000 < renderTime) {
```

Comment:
@mmocny this is the "windowing" solution I ended up going with, e.g. keep all entries in memory and drop them once they're more than 2 seconds old.

Let me know if you think 2 seconds is too long (or too short), or if you have a better suggestion to handle out-of-order dispatch.

Reply:
There is no single duration value that would guarantee anything, because some buffering can happen e.g. for as long as the user is holding down a finger.

Another way we could end up with >2s is just because the Performance timeline fires Observers with less than NORMAL priority tasks, and so may take a long time between observation calls-- even if the timeline data is much less than 2s apart.

---

However, because you are only doing this cleanup when new event timings arrive, I think this defends against a lot of the reasons this wouldn't work.

I think this [feature request](issues.chromium.org/issues/328902994) might also help resolve some of this.

---

I'll need more time to think through better alternatives, but this seems reasonable enough for most cases to me.  Any improvements in the future I don't think would be breaking API.

Reply:
> However, because you are only doing this cleanup when new event timings arrive, I think this defends against a lot of the reasons this wouldn't work.

Right, that was my hope. BTW, I've updated the implementation a bit to instead just keep track of the 10 most recently reported animation frame times and assume that any new event entry that is dispatched is not going to be more than 10 frames old.

If (for some reason) that turns out not to be true in some case, what would happen is some entries might not get properly attributed to their INP candidate, and I think that's an acceptable trade off (e.g. nothing will break, it's just won't be perfect attribution).